hal/src/nRF52840/device_code.cpp

```diff
@@ -38,29 +38,37 @@ const auto DEVICE_NAME_DCT_OFFSET = DCT_SSID_PREFIX_OFFSET;
 
 const auto DEVICE_NAME_MAX_SIZE = DEVICE_NAME_DCT_SIZE - 1;
 
-const auto SETUP_CODE_SIZE = DCT_DEVICE_CODE_SIZE;
 const auto SETUP_CODE_DCT_OFFSET = DCT_DEVICE_CODE_OFFSET;
 
-int getDeviceSetupCode(char* code) {
+} // ::
+
+int get_device_setup_code(char* code, size_t size) {
     // Check if the device setup code is initialized in the DCT
-    int ret = dct_read_app_data_copy(SETUP_CODE_DCT_OFFSET, code, SETUP_CODE_SIZE);
-    if (ret < 0 || !isPrintable(code, SETUP_CODE_SIZE)) {
+    char setupCode[HAL_SETUP_CODE_SIZE] = {};
+    int ret = dct_read_app_data_copy(SETUP_CODE_DCT_OFFSET, setupCode, sizeof(setupCode));
+    if (ret < 0 || !particle::isPrintable(setupCode, sizeof(setupCode))) {
         // Check the OTP memory
         char serial[HAL_DEVICE_SERIAL_NUMBER_SIZE] = {};
         ret = hal_get_device_serial_number(serial, sizeof(serial), nullptr);
-        if (ret >= (int)SETUP_CODE_SIZE && (size_t)ret <= sizeof(serial)) {
+        if ((size_t)ret >= sizeof(setupCode) && (size_t)ret <= sizeof(serial)) {
             // Use last characters of the serial number as the setup code
-            memcpy(code, &serial[ret - SETUP_CODE_SIZE], SETUP_CODE_SIZE);
+            size = std::min(size, (size_t)HAL_SETUP_CODE_SIZE);
+            memcpy(code, &serial[ret - size], size);
         } else {
             // Return a dummy setup code
-            memset(code, 'X', SETUP_CODE_SIZE);
+            size = std::min(sizeof(setupCode), size);
+            memset(code, 'X', size);
         }
+    } else {
+        size = std::min(sizeof(setupCode), size);
+        memcpy(code, setupCode, size);
     }
-    return 0;
+    if (size < sizeof(setupCode)) {
```

Comment:
Use a different variable to store the actual resulting size and keep `size` argument intact. Use that to infer whether we can null-terminate or not.